### PR TITLE
Some Netty performance boost stuff

### PR DESCRIPTION
**Pull request by xxDark**

### CODE REVIEW

File: proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java [hunk 1]
```diff
@@ -29,6 +30,7 @@
   private final EventLoopGroup bossGroup;
   private final EventLoopGroup workerGroup;
   private final VelocityServer server;
+  private final boolean epoll;
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java]: There's no need for this. Just see if `transportType == EPOLL`.

File: proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java [hunk 2]
```diff
@@ -75,6 +78,9 @@ public void bind(final InetSocketAddress address) {
         .childOption(ChannelOption.TCP_NODELAY, true)
         .childOption(ChannelOption.IP_TOS, 0x18)
         .localAddress(address);
+    if (this.epoll) {
+      bootstrap.option(EpollChannelOption.TCP_FASTOPEN, this.server.getConfiguration().getTcpFastOpenMode());
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java]: I totally understand why `TCP_FASTOPEN` should be enabled for client connections (from the proxy to the server), but I'm skeptical on why we should enable it for the server itself, given that Minecraft connections are supposed to be long-lived.

Additionally, I'm not aware of any Minecraft servers which have `TCP_FASTOPEN` on by default, as it's usually disabled. If we merge this in, we should give some instructions on how to make use of fast open on servers.

Comment (astei) [reply proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java]: Additionally, this line (and the others like it) will cause a Checkstyle error.

Comment (xxDark) [reply proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java]: >  
> supposed to be long-lived.
> 
Not all the players are playing fair.

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java [hunk 3]
```diff
@@ -94,6 +94,18 @@
   @Ignore
   private @Nullable Favicon favicon;
 
+  @Comment({"What TCP Fast Open mode should be used? (Works only if Epoll is available)",
+          "0 - disabled", "1 - TFO is only enabled for outgoing client",
+          "2 - TFO is only available on listening server",
+          "3 - includes 1 and 2"})
+  @ConfigKey("tcp-fast-open")
+  private int tcpFastOpenMode = 0;
```
Comment (Cybermaxke) [inline proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java]: Make this an enum?

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java [hunk 4]
```diff
@@ -94,6 +94,18 @@
...
+
+  @Comment({"Shuold port be reuseable for multiple proxy instances?",
+          "(Worls only if Epoll is available)"})
```
Comment (Crypnotic) [inline proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java]: Hate to be that guy, but in case you missed this. `Should` `Works`*

File: proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java (hunk 2 above)
Comment (astei) [reply proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java]: What do you mean by this?

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java (hunk 3 above)
Comment (astei) [reply proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java]: Yes, this definitely needs to be an enum.

File: proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java [hunk 5]
```diff
@@ -68,13 +69,19 @@ public void logChannelInformation() {
    */
   public void bind(final InetSocketAddress address) {
     final ServerBootstrap bootstrap = new ServerBootstrap()
-        .channel(this.transportType.serverSocketChannelClass)
+        .channelFactory(this.transportType.serverChannelFactory)
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java]: I'd like some hard numbers on this. How much do we gain by using a lambda channel factory? I'm not convinced we get much of a gain.

File: api/src/main/java/com/velocitypowered/api/event/EventManager.java [hunk 6]
```diff
@@ -57,9 +57,7 @@
    *
    * @param event the event to fire
    */
-  default void fireAndForget(Object event) {
-    fire(event);
-  }
+  void fireAndForget(Object event);
```
Comment (astei) [inline api/src/main/java/com/velocitypowered/api/event/EventManager.java]: This is an entirely separate change. (This also technically breaks the API, but seeing as `VelocityEventManager` is the only implementation we care about, that's fine.)

Please break this off into a separate PR.

File: proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java (hunk 5 above)
Comment (xxDark) [reply proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java]: Going to provide tomorrow/in next 2 days

Comment (xxDark) [reply proxy/src/main/java/com/velocitypowered/proxy/network/ConnectionManager.java]: Going to provide tomorrow/in next 2 days

**Comment by astei** [comment]
@xxDark Still interested in this? Rebase your PR on `dev/1.1.0` and respond to any outstanding comments. Otherwise I'll close this PR.

(Avoiding the creation of `CompletableFuture` for fire-and-forget events made it into Velocity already.)

**Comment by xxDark** [comment]
> 
> 
> @xxDark Still interested in this? Rebase your PR on `dev/1.1.0` and respond to any outstanding comments. Otherwise I'll close this PR.
> 
> (Avoiding the creation of `CompletableFuture` for fire-and-forget events made it into Velocity already.)

I don't think I can do this rn, guess you can just apply this changes by yourself if they are needed.

**Comment by astei** [comment]
Alright, sounds good. Sorry that you couldn't complete this PR.

**Comment by astei** [comment]
TCP fast open support is now in 1.1.0 (121ab41d9). It is best paired with a server that is modified to enable TCP fast open.
